src/libraries/System.Linq/src/System/Linq/ToCollection.cs

```diff
@@ -29,6 +29,18 @@ public static List<TSource> ToList<TSource>(this IEnumerable<TSource> source)
             return source is IIListProvider<TSource> listProvider ? listProvider.ToList() : new List<TSource>(source);
         }
 
+        public static Dictionary<TKey, TValue> ToDictionary<TKey, TValue>(this IEnumerable<KeyValuePair<TKey, TValue>> source) where TKey : notnull =>
+            source.ToDictionary(kv => kv.Key, kv => kv.Value, null);
```

Comment:
I wonder if the Dictionary constructor is faster in this case:

```suggestion
            new Dictionary(source);
```

Reply:
Hi @saul when I benchmark between them, I found constructor version is not always faster or lower allocation. Maybe the possible [reason](https://github.com/dotnet/runtime/blob/05a3d360d6079db3cb64636fcdf99f832a58c1d2/src/libraries/System.Linq/src/System/Linq/ToCollection.cs#L119) is Linq version specially cares some concrete collection types so that may: pre-allocate enough capacity (by ICollection.Count) and try to eliminate usage of Enumerable (eg array), which offsets delegate cost for those concrete collection types.
Data size: 100000
![image](https://user-images.githubusercontent.com/72254037/236769816-a5b54b79-5909-4a73-87f6-dd4f190e3c21.png)

Reply:
Can you share your benchmark?

Reply:
```csharp
[MemoryDiagnoser]
    public class TestClass1
    {
        private const int DataSize = 100000;
        private KeyValuePair<int, string>[] _kvs;
        private List<KeyValuePair<int, string>> _kvList;

        [GlobalSetup]
        public void Init()
        {
            _kvs = new KeyValuePair<int, string>[DataSize];

            for (int i = 0; i < _kvs.Length; i++)
            {
                _kvs[i] = new KeyValuePair<int, string>(i, string.Empty);
            }

            _kvList = new List<KeyValuePair<int, string>>(DataSize);
            for (int i = 0; i < DataSize; i++)
            {
                _kvList.Add(new KeyValuePair<int, string>(i, string.Empty));
            }
        }

        private static IEnumerable<KeyValuePair<int, string>> GeneratePureEnumerable() => Enumerable.Range(0, DataSize).Select(n => new KeyValuePair<int, string>(n, string.Empty));

        [Benchmark]
        public Dictionary<int, string> UseConstructor()
        {
            return new Dictionary<int, string>(_kvs);
        }
        
        [Benchmark]
        public Dictionary<int, string> UseLinq()
        {
            return _kvs.ToDictionary(kv => kv.Key, kv => kv.Value);
        }
        
        [Benchmark]
        public Dictionary<int, string> UseConstructorForList()
        {
            return new Dictionary<int, string>(_kvList);
        }
        
        [Benchmark]
        public Dictionary<int, string> UseLinqForList()
        {
            return _kvList.ToDictionary(kv => kv.Key, kv => kv.Value);
        }

        [Benchmark]
        public Dictionary<int, string> UseConstructorForSlowEnumerable()
        {
            return new Dictionary<int, string>(GeneratePureEnumerable());
        }

        [Benchmark]
        public Dictionary<int, string> UseLinqForSlowEnumerable()
        {
            return GeneratePureEnumerable().ToDictionary(kv => kv.Key, kv => kv.Value);
        }
    }
```

Reply:
Parameterization for different sizes would be useful. We pay attention for performance of small amount of data too.

Reply:
Data size: 10, 100, 1000, 100000. 
Note benchmark did not recognize List type so seems List related tests are not ordered. I overrides ToString() of pure enumerable wrapper for friendly reading in result.

// * Summary *

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22621
12th Gen Intel Core i7-12800H, 1 CPU, 20 logical and 14 physical cores
.NET SDK=7.0.105
  [Host]     : .NET 6.0.16 (6.0.1623.17311), X64 RyuJIT
  DefaultJob : .NET 6.0.16 (6.0.1623.17311), X64 RyuJIT


|                Method |         enumerable |                array |                list |           Mean |        Error |        StdDev |         Median |    Gen 0 |    Gen 1 |    Gen 2 |   Allocated |
|---------------------- |------------------- |--------------------- |-------------------- |---------------:|-------------:|--------------:|---------------:|---------:|---------:|---------:|------------:|
| CtorForSlowEnumerable | Enumerable[100000] |                    ? |                   ? | 3,006,546.9 ns | 60,089.89 ns | 156,181.61 ns | 3,007,032.0 ns | 468.7500 | 457.0313 | 453.1250 | 8,452,696 B |
| LinqForSlowEnumerable | Enumerable[100000] |                    ? |                   ? | 3,390,002.4 ns | 67,724.51 ns | 139,862.87 ns | 3,370,539.1 ns | 550.7813 | 539.0625 | 535.1563 | 8,452,764 B |
| CtorForSlowEnumerable |   Enumerable[1000] |                    ? |                   ? |    20,413.7 ns |  1,102.87 ns |   3,251.83 ns |    19,838.4 ns |   8.1177 |   2.6855 |        - |   102,272 B |
| LinqForSlowEnumerable |   Enumerable[1000] |                    ? |                   ? |    21,610.8 ns |  1,013.61 ns |   2,988.66 ns |    21,218.0 ns |   8.1177 |   2.6855 |        - |   102,272 B |
| CtorForSlowEnumerable |    Enumerable[100] |                    ? |                   ? |     2,249.4 ns |    104.80 ns |     309.02 ns |     2,239.4 ns |   0.8163 |   0.0362 |        - |    10,248 B |
| LinqForSlowEnumerable |    Enumerable[100] |                    ? |                   ? |     1,907.5 ns |     36.49 ns |      37.47 ns |     1,906.2 ns |   0.8163 |   0.0362 |        - |    10,248 B |
| CtorForSlowEnumerable |     Enumerable[10] |                    ? |                   ? |       258.4 ns |     13.77 ns |      40.59 ns |       256.0 ns |   0.0834 |   0.0002 |        - |     1,048 B |
| LinqForSlowEnumerable |     Enumerable[10] |                    ? |                   ? |       288.1 ns |     14.61 ns |      42.62 ns |       289.9 ns |   0.0834 |   0.0002 |        - |     1,048 B |
|          CtorForArray |                  ? | KeyVa(...)0000] [35] |                   ? | 1,314,024.8 ns | 26,208.33 ns |  57,527.92 ns | 1,309,202.3 ns | 244.1406 | 244.1406 | 244.1406 | 3,043,638 B |
|          LinqForArray |                  ? | KeyVa(...)0000] [35] |                   ? | 1,183,125.0 ns | 23,349.40 ns |  56,391.42 ns | 1,181,466.4 ns | 234.3750 | 234.3750 | 234.3750 | 3,042,666 B |
|          CtorForArray |                  ? | KeyVa(...)1000] [33] |                   ? |    10,435.0 ns |    686.07 ns |   2,022.89 ns |    10,172.0 ns |   2.4567 |   0.3967 |        - |    31,048 B |
|          LinqForArray |                  ? | KeyVa(...)1000] [33] |                   ? |     8,133.3 ns |    453.60 ns |   1,330.32 ns |     7,911.9 ns |   2.4567 |   0.4044 |        - |    31,016 B |
|          CtorForArray |                  ? | KeyVa(...)>[10] [31] |                   ? |       154.2 ns |      9.07 ns |      26.74 ns |       151.9 ns |   0.0376 |        - |        - |       472 B |
|          LinqForArray |                  ? | KeyVa(...)>[10] [31] |                   ? |       117.7 ns |      6.48 ns |      19.11 ns |       117.3 ns |   0.0350 |        - |        - |       440 B |
|          CtorForArray |                  ? | KeyVa(...)[100] [32] |                   ? |     1,122.8 ns |     65.03 ns |     191.73 ns |     1,064.6 ns |   0.2518 |   0.0057 |        - |     3,160 B |
|          LinqForArray |                  ? | KeyVa(...)[100] [32] |                   ? |       755.5 ns |     40.45 ns |     118.63 ns |       689.7 ns |   0.2489 |   0.0057 |        - |     3,128 B |
|           CtorForList |                  ? |                    ? | Syst(...)ng]] [104] |    11,599.5 ns |    196.73 ns |     184.02 ns |    11,602.2 ns |   2.4567 |   0.3967 |        - |    31,064 B |
|           CtorForList |                  ? |                    ? | Syst(...)ng]] [104] |     1,199.0 ns |     23.33 ns |      21.83 ns |     1,197.0 ns |   0.2518 |   0.0057 |        - |     3,176 B |
|           CtorForList |                  ? |                    ? | Syst(...)ng]] [104] |       158.7 ns |      3.00 ns |       2.80 ns |       159.0 ns |   0.0389 |        - |        - |       488 B |
|           CtorForList |                  ? |                    ? | Syst(...)ng]] [104] | 1,505,143.6 ns | 28,406.16 ns |  29,171.04 ns | 1,510,347.7 ns | 224.6094 | 224.6094 | 224.6094 | 3,043,303 B |
|           LinqForList |                  ? |                    ? | Syst(...)ng]] [104] |       896.1 ns |     17.73 ns |      22.42 ns |       891.0 ns |   0.2489 |   0.0057 |        - |     3,128 B |
|           LinqForList |                  ? |                    ? | Syst(...)ng]] [104] |    10,143.1 ns |    625.76 ns |   1,815.44 ns |     9,684.5 ns |   2.4567 |   0.3967 |        - |    31,016 B |
|           LinqForList |                  ? |                    ? | Syst(...)ng]] [104] | 1,352,849.5 ns | 29,902.82 ns |  87,699.79 ns | 1,342,968.5 ns | 257.8125 | 257.8125 | 257.8125 | 3,042,562 B |
|           LinqForList |                  ? |                    ? | Syst(...)ng]] [104] |       158.4 ns |      8.83 ns |      26.03 ns |       155.3 ns |   0.0350 |        - |        - |       440 B |

Reply:
Benchmark code:

```csharp
class NamedEnumerable : IEnumerable<KeyValuePair<int, string>>
    {
        private readonly int _dataSize;
        private readonly IEnumerable<KeyValuePair<int, string>> _underlyingEnumerable;

        public NamedEnumerable(int dataSize)
        {
            _dataSize = dataSize;
            _underlyingEnumerable = Enumerable.Range(0, dataSize).Select(i => new KeyValuePair<int, string>(i, string.Empty));
        }

        public IEnumerator<KeyValuePair<int, string>> GetEnumerator() => _underlyingEnumerable.GetEnumerator();

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }

        public override string ToString() => $"Enumerable[{_dataSize}]";
    }


    [MemoryDiagnoser]
    public class TestClass1
    {
        public IEnumerable<object> TestArrays() => TestPureEnumerables().Cast<IEnumerable<KeyValuePair<int, string>>>().Select(enumerable => enumerable.ToArray());

        public IEnumerable<object> TestLists() => TestPureEnumerables().Cast<IEnumerable<KeyValuePair<int, string>>>().Select(enumerable => enumerable.ToList());

        public IEnumerable<object> TestPureEnumerables()
        {
            yield return new NamedEnumerable(10);
            yield return new NamedEnumerable(100);
            yield return new NamedEnumerable(1000);
            yield return new NamedEnumerable(100000);
        }

        [Benchmark]
        [ArgumentsSource(nameof(TestArrays))]
        public Dictionary<int, string> CtorForArray(KeyValuePair<int, string>[] array)
        {
            return new Dictionary<int, string>(array);
        }
        
        [Benchmark]
        [ArgumentsSource(nameof(TestArrays))]
        public Dictionary<int, string> LinqForArray(KeyValuePair<int, string>[] array)
        {
            return array.ToDictionary(kv => kv.Key, kv => kv.Value);
        }
        
        [Benchmark]
        [ArgumentsSource(nameof(TestLists))]
        public Dictionary<int, string> CtorForList(List<KeyValuePair<int, string>> list)
        {
            return new Dictionary<int, string>(list);
        }
        
        [Benchmark]
        [ArgumentsSource(nameof(TestLists))]
        public Dictionary<int, string> LinqForList(List<KeyValuePair<int, string>> list)
        {
            return list.ToDictionary(kv => kv.Key, kv => kv.Value);
        }
        
        [Benchmark]
        [ArgumentsSource(nameof(TestPureEnumerables))]
        public Dictionary<int, string> CtorForSlowEnumerable(IEnumerable<KeyValuePair<int, string>> enumerable)
        {
            return new Dictionary<int, string>(enumerable);
        }
        
        [Benchmark]
        [ArgumentsSource(nameof(TestPureEnumerables))]
        public Dictionary<int, string> LinqForSlowEnumerable(IEnumerable<KeyValuePair<int, string>> enumerable)
        {
            return enumerable.ToDictionary(kv => kv.Key, kv => kv.Value);
        }
    }
```

Reply:
Fixed.